src/ruby/travelling-salesman.rb

```diff
@@ -27,81 +26,55 @@ def shortestPath
 
     # add option to current order and distance
     private
-    def addOption(distance, node)
-        nodeIndex = @nodes.index(node)
-        for row in @matrix 
-            row[nodeIndex] = 0
-        end
+    def addOption(distance, index)
+        @matrix.each {|row| row[index] = 0}
 
-        @order << node
+        @order << @nodes[index]
         @distance += distance
     end
 
     # goes to next node to add to route
-    private
-    def addNextNode(distance, node)
-        nodeIndex = @nodes.index(node)
-
-        if (noOptionsAvailable(@matrix[nodeIndex]))
+    def addNextNode(distance, index)
+        if (noOptionsAvailable(@matrix[index]))
             updateShortest
-            goBackNode(distance, node)
+            goBackNode(distance, index)
         else
-            for option in @matrix[nodeIndex]
-                nextNodeIndex = @matrix[nodeIndex].index(option)
-                nextNode = @nodes[nextNodeIndex]
-                if (option != 0)
-                    addOption(option, nextNode)
-                    addNextNode(option, nextNode)
-                end
+            @matrix[index].each_with_index do |option, index|
+              if (option != 0)
+                  addOption(option, index)
+                  addNextNode(option, index)
+              end
             end
-
-            goBackNode(distance, node)
+            goBackNode(distance, index)
         end
     end
 
     # check to see if at a dead end
-    private
     def noOptionsAvailable(row)
-        deadEnd = true
-        for cell in row
-            if( cell != 0)
-                deadEnd = false
-            end
-        end
-
-        return deadEnd
+        row.none? {|cell| cell != 0}
     end
 
     # go back to previous node to go down other paths from that node
-    private
-    def goBackNode(distance, node)
+    def goBackNode(distance, nodeIndex)
         original  = createStartMatrix
-        nodeIndex = @nodes.index(node)
-
-        i=0
-        while i < original.length
-            @matrix[i][nodeIndex] = original[i][nodeIndex]
-            i += 1
-        end
+        original.length.times {|i| @matrix[i][nodeIndex] = original[i][nodeIndex]}
```

Comment:
Chaining methods is one of the really lovely things about the language, it makes it super readable too. Because numbers in Ruby are objects, you can just call times on the result of length and not have to worry about creating a counter :)